### ISSUE TITLE
DicomSplitter

### ISSUE THREAD

**Issue by asvcode** [fastai/fastai#2724]
**Feature Request.**
Dicom images provide unique nuggets of information that can help in ensuring that the same patient does not exist in both the train and valid sets.  Having the same patient in both sets may potentially skew the effectiveness of the model.  `DicomSplitter` utilizes `IndexSplitter` which splits items so that `val_idx` is in the validation set and the others in the training set.

**Proposed Method**
```
def DicomSplit(valid_pct=0.2, seed=None, **kwargs):
    "Splits `items` between train/val with `valid_pct`"
    "and checks if identical patient IDs exist in both the train and valid sets"
    def _inner(o, **kwargs):
        train_list=[]; valid_list=[]
        if seed is not None: torch.manual_seed(seed)
        rand_idx = L(int(i) for i in torch.randperm(len(o)))
        cut = int(valid_pct * len(o))
        trn = rand_idx[cut:]; trn_p = o[rand_idx[cut:]]
        val = rand_idx[:cut]; val_p = o[rand_idx[:cut]]
        
        for i, im in enumerate(trn_p):
            trn = im.dcmread()
            patient_ID = trn.PatientID
            train_list.append(patient_ID)
        for j, jm in enumerate(val_p):
            val = jm.dcmread()
            vpatient_ID = val.PatientID
            valid_list.append(vpatient_ID) 
        print(f'Duplicate: {set(train_list) & set(valid_list)}')
        return rand_idx[cut:], rand_idx[:cut]
    return _inner
```

`DicomSplit` creates a train and valid list and checks to see if there are any duplicate patient IDs between the two sets.

`DicomSplitter` is an adaptation of `IndexSplitter` which takes into consideration the train and valid split formed by `DicomSplit`

```
def DicomSplitter(items, valid_pct=0.2):
    trn, val = DicomSplit(valid_pct=valid_pct)(items)
    valid_idx = val
    def _inner(o):
        train_idx = np.setdiff1d(np.array(range_of(o)), np.array(valid_idx))
        print(f'train:{train_idx} val:{valid_idx}')
        return L(train_idx, use_list=True), L(valid_idx, use_list=True)
    return _inner
```

`DicomSplitter` can then be used with datablock via the `splitter` tab

```
melanoma = DataBlock(blocks=blocks,
                   get_x=get_x,
                   splitter=DicomSplitter(mel_items_500),
                   item_tfms=Resize(196),
                   get_y=ColReader('target'),
                   batch_tfms=batch_tfms)

dls = melanoma.dataloaders(df_mel, bs=64, num_workers=0)
dls.show_batch()
```

**Example**
Dataset used is from this [kaggle competition](https://www.kaggle.com/c/siim-isic-melanoma-classification/data)

```
source = 'D:/Datasets/Melanoma/'
files = os.listdir(source)
mel_items = get_dicom_files(source, folders='train')
```
This is a large dataset so lets create a random smaller version (500 images) for testing
```
randomList = []
for i in range(0, 500):
    randomList.append(random.randint(0, 10000))

mel_items_500 = mel_items[randomList]
```
> Sanity Check 
seed = 7
```
def check_duplicate(items, seed=7):
    trn, val = DicomSplit(valid_pct=0.2, seed=seed)(mel_items_500)
    print(trn, val)
```
`check_duplicate(mel_items_500)`
![sanity](https://user-images.githubusercontent.com/25020209/91652245-0fb1a000-ea4a-11ea-99a3-6cee253ae23c.PNG)

seed = 77

![sanity2](https://user-images.githubusercontent.com/25020209/91652373-8b601c80-ea4b-11ea-845d-c931a870b99a.PNG)

> Datablock
```
get_x = lambda x:f'{source}train/{x[0]}.dcm'
get_y=ColReader('target')
batch_tfms = aug_transforms(flip_vert=True, max_lighting=0.1, max_zoom=1.05, max_warp=0.)
blocks = (ImageBlock(cls=PILDicom), CategoryBlock)
```

```
melanoma = DataBlock(blocks=blocks,
                   get_x=get_x,
                   splitter=DicomSplitter(mel_items_50, seed=7),
                   item_tfms=Resize(196),
                   get_y=ColReader('target'),
                   batch_tfms=batch_tfms)

dls = melanoma.dataloaders(df_mel, bs=64, num_workers=0)
dls.show_batch()
```
![sanity1](https://user-images.githubusercontent.com/25020209/91652286-99616d80-ea4a-11ea-95d1-b1bb2bfa9aee.PNG)

In this case there is 1 duplicate patient ID

**Comment by antorsae** [comment]
If I understand code correctly the above would read all dicoms and attempt a split.

Given that fastai already provides a convenient (and paralellized pd.DataFrame.from_dicoms(...) ) method, I suggest instead to create a generic (i.e. not dicom-specific) splitter:

```
def ColGroupKFoldSplitter(col,n_folds:int=5,fold:int=0):
    "Split `items` (supposed to be a dataframe) by GroupKFold in`col`"
    def _inner(o):
        assert isinstance(o, pd.DataFrame), "ColSplitter only works when your items are a pandas DataFrame"
        # TODO
    return _inner
```

and then assuming you're instanciating the DataBlock on df_mel, you'd just do: `ColGroupKFoldSplitter('PatientID')`
